### ISSUE TITLE
Allow deferred template changes

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#8070]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

Currently, the Qube Manager will not allow you to change a running qube's template. (The field is grayed out.) The qube must be shut down first.

In addition, if you try to change the template on which a disposable template is based while there are running disposables based on that disposable template, you will receive an error message stating that the action cannot be performed due to the running disposables.

### The solution you'd like

As @marmarek suggested in https://github.com/QubesOS/qubes-doc/pull/1297#issuecomment-1454343087, it should be possible for the user to change the template in the GUI even when qubes based on that template are running, but defer the actual change until after those running qubes shut down or reboot.

### The value to a user, and who that user might be

Users will be able to tell the system what to do at any time, and the system will more intelligently handle those directives at the appropriate time, relieving users of some micromanagement.

**Comment by DemiMarie** [comment]
I’ll start work tomorrow.

**Comment by DemiMarie** [comment]
Some thoughts on this issue:

1. Using a feature to change the template is _very_ counterintuitive, not least because `qvm-prefs` _is_ the most common way to change the TemplateVM from the command line.  Furthermore, as mentioned above this breaks qrexec policy for the Admin API.  The simplest solution to both problems I can think of is for the `template` property to always refer to the _pending_ value, and use a different property (`old_template`?) for the stale value.
2. For normal disposables, would it be possible to leave the template dangling?

**Comment by marmarek** [comment]
> Using a feature to change the template is _very_ counterintuitive, not least because `qvm-prefs` _is_ the most common way to change the TemplateVM from the command line. Furthermore, as mentioned above this breaks qrexec policy for the Admin API. The simplest solution to both problems I can think of is for the `template` property to always refer to the _pending_ value, and use a different property (`old_template`?) for the stale value.

The "current" value definitely is needed somewhere. If nothing else, it's relevant for example for updates check (if app qube reports updates available, it needs to be marked on the correct template), but likely several more things. I wouldn't call it "old_template", as it isn't really old - it's the current one. What you propose is more or less the same as I described but using a property not a feature. I guess that's fine too. Maybe "template" and "current_template" or "template_current" (the latter being read-only)?

> For normal disposables, would it be possible to leave the template dangling?

It still need to keep info with what template it was started.

**Comment by DemiMarie** [comment]
> > Using a feature to change the template is _very_ counterintuitive, not least because `qvm-prefs` _is_ the most common way to change the TemplateVM from the command line. Furthermore, as mentioned above this breaks qrexec policy for the Admin API. The simplest solution to both problems I can think of is for the `template` property to always refer to the _pending_ value, and use a different property (`old_template`?) for the stale value.
> 
> The "current" value definitely is needed somewhere. If nothing else, it's relevant for example for updates check (if app qube reports updates available, it needs to be marked on the correct template), but likely several more things. I wouldn't call it "old_template", as it isn't really old - it's the current one. What you propose is more or less the same as I described but using a property not a feature. I guess that's fine too. Maybe "template" and "current_template" or "template_current" (the latter being read-only)?

I currently have writes to `template` magically be redirected to `pending_template`, which cannot be written directly.  The changes take effect the next time the VM shuts down, or when qubesd starts up while the VM has been shut down.  This only fails if VMs can start without the intervention of qubesd, which should not be the case.

**Comment by marmarek** [comment]
I'm not a fan on write redirection (I have considered it with my "feature" idea, and also rejected that). It means reading the property after writing it will return the old value, even when setting it worked. I don't want to have special cases for this property all over the code base.

**Comment by marmarta** [comment]
Some things that need to be figured out (those are not just UX, there is a lot of backend logic here):
- how this interacts with applications tab in VM settings - which apps are shown? how to make it not confusing? maybe disable the tab completely if there is a template change pending (in a similar way firewall can be disabled - with an explicit message in the tab)
- everything related to "check with template" stuff in VM settings, global config etc. - what if supported services change? which ones should we use? from the new one? 
- how to show it clearly and concisely and without messing up sorting/filtering in qube manager
- how to show it clearly in Template Switcher if you have previously changed the template with template switcher already, but haven't restarted the VM
- this VM probably should also be marked as stale/restart required in qube manager and domains widget


Not gonna lie, my first impression is that I hate it and the template change should not need a restart at all.


In places like qvm-ls and qvm prefs I think we can just show template: new-template (* pending change from old-template), we can add this information to gui vm settings too, but I have no good idea how to show it in qube manager and template switcher.

**Comment by marmarta** [comment]
> I'm not a fan on write redirection (I have considered it with my "feature" idea, and also rejected that). It means reading the property after writing it will return the old value, even when setting it worked. I don't want to have special cases for this property all over the code base.

everything that shows the template or uses check_with_template will have special cases :/

**Comment by ben-grande** [comment]
> - Decide on a "feature" name that can be set to change the template at the next possible time. For example `change-template`

Sounds like a verb, question that would require boolean response. I suggest `template_to_assign` if a property, else change the underline for dashes if a feature.

> - Implement changing the template at the next possible time. It should consider shutting down the VM itself, shutting down any disposable based on a disposable template with that feature set. But also, as a last resort thing, next system startup, and maybe also before starting the VM.

Makes sense to use `property-load` as Demi used, for the shutdown, I'd use `domain-shutdown` or `domain-post-shutdown` (to be done if `domain-shutdown` is too early in the process (DispVMs may still be killing the qube here, possibly irrelevant as this seems the case only for unnamed disposable, didn't dig further yet).

> - Guard setting the feature with similar checks as setting the template itself. This includes checking if the template exists, but also firing related `admin-permission` as for setting the template itself. Not sure about evaluating also qrexec policy for `admin.vm.property.Set+template` (doing so will avoid simple policy bypass if somebody forgets denying the feature set too, but might be also too invasive)

Seems like a hard one, there is no *dry-run* mode, there is `qrexec-policy --just-evaluate`, but didn't test if this is available on GUI domain. A not so nice solution though.

> - Handle the case when template can be set immediately - should it just do it? or maybe should it refuse to set the feature and ask the user to set the property directly?

When can be set immediately, yes, it should change. I don't like refusing the call if it can be set directly, consider that we want idempotent commands, would be nice to have it never fail a Salt/Ansible state that does not do error handling.

> - Consider the case of DispVM template - for normal disposables, it doesn't really make sense as they will be removed on shutdown; but maybe makes sense for named disposables?

Makes sense for named disposables.

> - Correctly handle new disposable created when its template has the feature set already (that feature will point at normal template, not disposable template then). Maybe simply avoid copying this feature when creating disposable

Agree with both.

> - Adjust user interface to use this feature when changing template, but also to show there is pending template change. This includes:
>  - qube settings
>   - domains widget (the tooltip)
>   - template switcher
>   - qube manager (template column)
>   - qvm-prefs (?)
>   - qvm-ls (?)

I believe that every tool must be consistent and show the same information, but I think some concession can be made to GUI tools, as I wrote at the end.

1. The current template listed on the same place it already is on each tool
2. For GUI tools, a warning icon with tooltip or warning text showing there is a deferred template change. For CLI tools, this is harder, people parse the output, it is not correct to add more information to the template field, `qvm-ls` should use a new column and `qvm-prefs` should have it's own separate key for storing the deferred template (if use property and not feature).

> Some things that need to be figured out (those are not just UX, there is a lot of backend logic here):
> - how this interacts with applications tab in VM settings - which apps are shown? how to make it not confusing? maybe disable the tab completely if there is a template change pending (in a similar way firewall can be disabled - with an explicit message in the tab)
> - everything related to "check with template" stuff in VM settings, global config etc. - what if supported services change? which ones should we use? from the new one? 
> - how to show it clearly and concisely and without messing up sorting/filtering in qube manager
> - how to show it clearly in Template Switcher if you have previously changed the template with template switcher already, but haven't restarted the VM
> - this VM probably should also be marked as stale/restart required in qube manager and domains widget
> 
> 
> Not gonna lie, my first impression is that I hate it and the template change should not need a restart at all.
> 
> 
> In places like qvm-ls and qvm prefs I think we can just show template: new-template (* pending change from old-template), we can add this information to gui vm settings too, but I have no good idea how to show it in qube manager and template switcher.

From what I read on this issue, reporting the current template instead of the deferred template is better as it avoids some of the issues described above, at least on the backend when reporting `qvm-prefs QUBE template`.

Creating a discrepancy of what the user see on different tools can cause confusion but maybe, can be done in the following scenario:

1. The property always reports the current template in use, so CLI tools should always be correct when gathering the property itself, but can read the deferred change via another property/feature. This is better for `check_with_template`, Qrexec policies, pending updates, app-menu tree view and everything else that uses this property.
2. GUI tools can use what you proposed, show the deferred template on the template field with a warning/asterisk informing that it is a pending change and the current template name.

**Comment by marmarta** [comment]
> Sounds like a verb, question that would require boolean response. I suggest `template_to_assign` if a property, else change the underline for dashes if a feature.

Maybe `next_template` or `future_template`?



> For GUI tools, a warning icon with tooltip or warning text showing there is a deferred template change. For CLI tools, this is harder, people parse the output, it is not correct to add more information to the template field, `qvm-ls` should use a new column and `qvm-prefs` should have it's own separate key for storing the deferred template (if use property and not feature).

I don't agree on `qvm-ls` using only the new column; something like old (new) or new (old) is better, because otherwise you get a different result from `qvm-ls` and from using your eyes to see what is available in the appvm... This is generally my issue with qube manager too - I think both need to be somehow reported in the field, a tooltip is acceptable additional info but it shouldn't be the only way to understand the column.

**Comment by marmarek** [comment]
> I don't agree on `qvm-ls` using only the new column; something like old (new) or new (old) is better, because otherwise you get a different result from `qvm-ls` and from using your eyes to see what is available in the appvm...

I agree both values should be visible somehow, but the issue is that people do use scripts to parse output of qvm-prefs/qvm-ls, and changing "template" to "template (something)" will break it. At least qvm-ls has an option for machine-readable output, so we can distinguish cases where it's for a human and when it's for a machine. Maybe it could be implemented by doing "smart template" column to be used for a human, while keeping simple template (and another column for future template) for machine? But we don't have that luxury with qvm-prefs...

**Comment by unman** [comment]
On Thu, Jul 03, 2025 at 04:47:27AM -0700, Marta Marczykowska-Górecka wrote:
> marmarta left a comment (QubesOS/qubes-issues#8070)
> Maybe `next_template` or `future_template`?
I think template_to_assign is clearer than both of these.
> 
> I don't agree on `qvm-ls` using only the new column; something like old (new) or new (old) is better, because otherwise you get a different result from `qvm-ls` and from using your eyes to see what is available in the appvm... This is generally my issue with qube manager too - I think both need to be somehow reported in the field, a tooltip is acceptable additional info but it shouldn't be the only way to understand the column.

I think that Ben intended that `qvm-ls` should use a new column in its
output, not that it should only use the new template_to_assign data. I
agree with his suggestion (if I understand it right). This will not
break parsing based on current output, but will be clear at the CLI.

**Comment by HW42** [comment]
> > Sounds like a verb, question that would require boolean response. I suggest `template_to_assign` if a property, else change the underline for dashes if a feature.
> 
> Maybe `next_template` or `future_template`?

FWIW: To me "pending" sounds better than next/future/to_assign (don't care about the order (`template_${term}` vs `${term}_template`))

> > I don't agree on `qvm-ls` using only the new column; something like old (new) or new (old) is better, because otherwise you get a different result from `qvm-ls` and from using your eyes to see what is available in the appvm...
> 
> I agree both values should be visible somehow, but the issue is that people do use scripts to parse output of qvm-prefs/qvm-ls, and changing "template" to "template (something)" will break it. At least qvm-ls has an option for machine-readable output, so we can distinguish cases where it's for a human and when it's for a machine. Maybe it could be implemented by doing "smart template" column to be used for a human, while keeping simple template (and another column for future template) for machine? But we don't have that luxury with qvm-prefs...

I think most cases should be addressable by treating this as a special additional property. In most cases you can just add that property to the list of things to show by default.

Only `qvm-prefs $vm template` stays tricky. **Maybe** print to stderr a warning, **maybe** only if it's a terminal (Really not sure about this part. Maybe just assume people using it are familiar with it enough).

**Comment by marmarta** [comment]
WRT https://github.com/QubesOS/qubes-issues/issues/7015 - can we do the same thing for other properties you can't change on a running VM, like color?

**Comment by ben-grande** [comment]
> I think that Ben intended that `qvm-ls` should use a new column in its output, not that it should only use the new template_to_assign data.

Right.

> Only `qvm-prefs $vm template` stays tricky. **Maybe** print to stderr a warning, **maybe** only if it's a terminal (Really not sure about this part. Maybe just assume people using it are familiar with it enough).

I like the stderr idea.

> WRT [#7015](https://github.com/QubesOS/qubes-issues/issues/7015) - can we do the same thing for other properties you can't change on a running VM, like color?

It can be done. One difference I see is that color could be changed change on runtime if GUI daemon support is added while deferred templates cannot because it is a drastic change to the system. Could consider that deferred color is a temporary workaround, it would [greatly benefit other cases (9810)](https://github.com/QubesOS/qubes-issues/issues/9810) if it was implemented on the GUI daemon side.

**Comment by tlaurion** [comment]
Any update here?

**Comment by ben-grande** [comment]
> Some things that need to be figured out (those are not just UX, there is a lot of backend logic here):
> 
>     * how this interacts with applications tab in VM settings - which apps are shown? how to make it not confusing? maybe disable the tab completely if there is a template change pending (in a similar way firewall can be disabled - with an explicit message in the tab)

I think it should display information and refresh from the old template. We don't know for how long the user will have the qube running until the next restart, so I don't think that we should disable the application tab. They might want to edit application listing.

>     * everything related to "check with template" stuff in VM settings, global config etc. - what if supported services change? which ones should we use? from the new one?

Old. This impacts service calls, must always be with the old template, as it's still in use.

>     * how to show it clearly and concisely and without messing up sorting/filtering in qube manager

I think it should display the template in use with an asterisk/warning-icon with tooltip indicating the pending one. Unless. we add a new column for the deferred template.

>     * how to show it clearly in Template Switcher if you have previously changed the template with template switcher already, but haven't restarted the VM

In the case of the switcher, either new column or substitute the value in the current column. I know, this contradicts what I said for the manager, but I think the switcher is a different case, it's not supposed to show the template being currently used, but the template that is set. Anyway, if there is disagreement about consistency, I plan to use a new column. I am less worried on using a new column here, because the Switcher window has space, while the manager, already has many columns.

>     * this VM probably should also be marked as stale/restart required in qube manager and domains widget

Yes.

---

> > * Correctly handle new disposable created when its template has the feature set already (that feature will point at normal template, not disposable template then). Maybe simply avoid copying this feature when creating disposable
> 
> Agree with both.

I agreed in the past and still agree now, but this one is harder than the rest. Things get much funnier when we add more template layers, used by disposables:

2 layers:

- template -> disposable template -> disposable (running): change disposable template: direct descendant has deferred template
- template -> appvm (running): change template: direct descendant has deferred template
  - always query current template being used
  - set `template` on appvm shutdown
  - "easy"

3 layers:

- template -> disposable template -> disposable (running): disposable template has deferred template
  - creating a new disposable, should use the deferred template, so it should use the deferred template from disposable template
  -  For querying features with `dispvm.check_with_template(...)` to work, `qubes.features.Features._recursive_check` will need to use the deferred value of `template` from `disposable template`, which, for the new disposable, is not deferred at all. Except, we might use the method from the next item.
  - For properties that are set doing the instantiation of the class, that uses `qubes.vm.qubesvm._default_with_template`, such as `memory`, `maxmem`, `vcpus`, `kernel`, `default_user`, `qrexec_timeout`, `shutdown_timeout`, `management_dispvm`, let's say the disposable template was loaded with those properties, but the new template has a different set, the new disposable will need to instantiate a dummy disposable template object to use the correct attributes (features etc).
  - Somehow, two disposables descendants from the same disposable template, will have different disposable template objects and will branch to different templates, beautiful drawing below:

When first reading this issue, I imagined this:

- dispA (old prior to deferred template) -> disposable template (default-dvm) -> template (fedora)
- dispB (new post deferred template) -> disposable template (default-dvm) (same object as above)-> template (debian)

But now, I think we will need dummy qube instances:

- dispA (old prior to deferred template) -> disposable template (default-dvm) -> template (fedora)
- dispB (new post deferred template)  -> disposable template (default-dvm) (different object than above) -> template (debian)

But this different object creates some problems. If it has the same name or qid, when attempting to load `self.app.domains`, which object would you want to get... Also another problem is, what if I notice that `dispB` is missing some file that should in `/rw/config/rc.local`, owned by the disposable template, and I want to start the disposable template to customize it, what if the other instance is already running.

Or we might prohibit changing template of disposable template in case the disposable template is running, as well as preventing the start of the disposable template, while still allowing.

Or a dummy object is a bad idea, but I still decided to keep it mentioned on this thread to evaluate all options, and I'd be glad if more ideas could be shared.

Edit: Marek mentioned to look at the root volume to get the correct inheritance.

**Comment by marmarta** [comment]
> I think it should display information and refresh from the old template. We don't know for how long the user will have the qube running until the next restart, so I don't think that we should disable the application tab. They might want to edit application listing.

I don't think it should refresh from the old template. It definitely needs to clearly communicate what is happening, and I don't think it should allow you to make changes based on the old template which would fail to work with the new template. You don't expect to be able to make changes that don't make sense, and this is a pretty senseless change. And we do want to encourage eventually restarting the qube to have the new template "on". 



> Old. This impacts service calls, must always be with the old template, as it's still in use.

Again, I don't like a situation in which you may unknowingly make changes that stop working after reboot. That's not a good choice to just "use old". At the very least it needs to be clearly communicated everywhere, but I don't like sticking red labels everywhere as a solution.... 



> I think it should display the template in use with an asterisk/warning-icon with tooltip indicating the pending one. Unless. we add a new column for the deferred template.

Not a new column, definitely no. Asterisk maybe, but I don't think it should only be in a tooltip, maybe something like an exclamation mark icon and something like `current_template \n (new template)`



> In the case of the switcher, either new column or substitute the value in the current column. I know, this contradicts what I said for the manager, but I think the switcher is a different case, it's not supposed to show the template being currently used, but the template that is set. Anyway, if there is disagreement about consistency, I plan to use a new column. I am less worried on using a new column here, because the Switcher window has space, while the manager, already has many columns.

no on a new column, and I think it should also show both values, current and pending one.

**Comment by ben-grande** [comment]
> > > ```
> > > * everything related to "check with template" stuff in VM settings, global config etc. - what if supported services change? which ones should we use? from the new one?
> > > ```
> > 
> > 
> > Old. This impacts service calls, must always be with the old template, as it's still in use.
> 
> Again, I don't like a situation in which you may unknowingly make changes that stop working after reboot. That's not a good choice to just "use old". At the very least it needs to be clearly communicated everywhere, but I don't like sticking red labels everywhere as a solution.... 

Well, this is a bit harder than what I imagined, it is not absolute that we should do one or the other.

In case of bootmode, we want to get it from the new template: https://github.com/QubesOS/qubes-manager/blob/b2fea1d2761418abde7f745fd25c5780fb052467/qubesmanager/settings.py#L66.

But for firewall, we want to get it from the current template: https://github.com/QubesOS/qubes-manager/blob/b2fea1d2761418abde7f745fd25c5780fb052467/qubesmanager/settings.py#L455 to not lose the firewall tab while the qube is still running.

**Comment by ben-grande** [comment]
> > I'm not a fan on write redirection (I have considered it with my "feature" idea, and also rejected that). It means reading the property after writing it will return the old value, even when setting it worked. I don't want to have special cases for this property all over the code base.
> 
> everything that shows the template or uses check_with_template will have special cases :/

I see the value in both ideas, redirection or or replacement. Properties that can't change in-qube after startup, such as `kernel`, `memory` and `stubdom_mem`, (important for non-memory balanced domains), `virt_mode`, `bootmode`, `default_user` (maybe breaks Qrexec calls), `guivm` (I don't think switching works right now), aren't problematic. The `template` though, that is used on the server side for querying properties, features

For now, I am implementing with replacement, where changing a template while the qube is running, does the following:

- `template`: new value
- `active_template`: old value

The default of `active_template` is `template`.

When we want to gather information from the template, we want to know what will work on the currently running / active system, therefore, the queries that were made to `template`, where appropriate, will be made to `active_template`. At least on the server side, from the changes I made (not pushed yet), it works (tested only AppVMs, not DVMTemplateMixin and DispVM).

> > I think it should display information and refresh from the old template. We don't know for how long the user will have the qube running until the next restart, so I don't think that we should disable the application tab. They might want to edit application listing.
> 
> I don't think it should refresh from the old template. It definitely needs to clearly communicate what is happening, and I don't think it should allow you to make changes based on the old template which would fail to work with the new template. You don't expect to be able to make changes that don't make sense, and this is a pretty senseless change. And we do want to encourage eventually restarting the qube to have the new template "on".

Can you give examples? For example, when we check feature against a template, it is to set a default (most of the times). If the user sets a value, they are overriding the default, so the change will make sense next boot. The only thing that is worrying me a bit is if the current features in the template restricts the available values to choose from.

> In case of bootmode, we want to get it from the new template: https://github.com/QubesOS/qubes-manager/blob/b2fea1d2761418abde7f745fd25c5780fb052467/qubesmanager/settings.py#L66.

I want to rectify, I think this is ok to be fetched from `active_template`. But of course, would need to test how the clients, such as Qube Manager choosing boot mode, would react.

**Comment by ben-grande** [comment]
For interested parties, please take a look on the server side implementation:

- https://github.com/QubesOS/qubes-core-admin/pull/870

**Comment by ben-grande** [comment]
After doing the server, it is easier to think what the clients should do:

- Template Switcher:
  - No need to change the "Current" column or add any, it already reflects the `template` property and that is what matter when switching
  - Disable the running check that disables the radio list and shows "_qube is running_"

- QUI domains:
  - Monitor also `property-(re)?set:active_template`
  - When `active_template` is not the default:
    - Add the outdated marker
    - Show the value of `template` and `active_template` in the tooltip. This makes sense to indicate why the qube is outdated when it hasn't even started and the volumes haven't changed yet. The tooltip already has a message "_Restart qube to apply changes in template._"

- Global Config
  - No event handling exists
  - On places that query the `template` property, change it to `active_template`. Only found it relevant on the "_Disposables_" tab, "_BASED ON_" column. I think the `active_template` is better here because if you enable preloading, their ascendant will be the `active_template` of the disposable template.

- Qube Manager

    > > I think it should display the template in use with an asterisk/warning-icon with tooltip indicating the pending one. Unless. we add a new column for the deferred template.
    > 
    > Not a new column, definitely no. Asterisk maybe, but I don't think it should only be in a tooltip, maybe something like an exclamation mark icon and something like current_template \n (new template)

  - Besides what is quote above, right-click context menu should be enabled for running qubes

- Qube Settings
  - There is a "_(current)_" marker that shows the value of `template`. I think it should continue showing that. No need to show `active` template in the radio list itself.
  - I think it makes sense to have a warning on the right side of the radio list (outside of it), that when the qube is running and has non-default `active_template`, show its value indicating what is currently in use.
  - For the applications tab, I still think we shouldn't disable it because we don't know for how long the user might continue running the qube, after changing the template, and would still like to refresh the application listing. We can discuss this later.

- qvm-prefs
  - No code changes required. Here is what it shows right now:
    - Default:
    ```
    active_template       D  fedora-44-xfce
    template              -  fedora-44-xfce
    ```
    - Change the template while qube is running:
    ```
    active_template       -  fedora-44-xfce
    template              -  archlinux
    ```
  - Requires documenting the `active_template` property.

- qvm-ls:

    ```sh
    % qvm-ls test-def-destroy
    NAME              STATE   CLASS  LABEL  TEMPLATE   NETVM
    test-def-destroy  Halted  AppVM  red    archlinux  sys-firewall

    % qvm-ls test-def-destroy --fields name,template,active_template
    NAME              TEMPLATE   ACTIVE-TEMPLATE
    test-def-destroy  archlinux  archlinux

    % qvm-ls --raw-data test-def-destroy --fields name,template,active_template
    test-def-destroy|archlinux|archlinux
    ```

    - Add the new column `ACTIVE-TEMPLATE` to the these formats:
    ```
    % qvm-ls --help-formats | grep -i template
      kernel   NAME,STATE,CLASS,TEMPLATE,KERNEL,KERNELOPTS
      prefs    NAME,LABEL,TEMPLATE,NETVM,VCPUS,INITIALMEM,MAXMEM,VIRT_MODE
      simple   NAME,STATE,CLASS,LABEL,TEMPLATE,NETVM
   ```
